### PR TITLE
Fix arm shit including some Hacks. But without this PR, robot is useless

### PR DESCRIPTION
**Pull request by MatthijsBurgh**
(Also fixes GPSR)

Requires https://github.com/tue-robotics/action_server/pull/112

### CODE REVIEW

File: challenge_gpsr/src/gpsr.py [hunk 1]
```diff
@@ -97,14 +97,13 @@ def _on_request_missing_information(self, description, grammar, target):
...
 
-
                 if not self.is_text_valid_input(sentence):
                     self._say_to_user("I don't understand what you're saying, please rephrase")
                     continue
 
                 self.timeout_count = 0
                 correct = True
-                if self.test:
+                if not self.test:
```
Comment (LarsJanssenTUe) [inline challenge_gpsr/src/gpsr.py]: Why has this changed?

File: robot_skills/src/robot_skills/arms.py [hunk 2]
```diff
@@ -126,6 +119,12 @@ def send_joint_trajectory(self, configuration, timeout=5):
                        "Specify get_arm(..., required_trajectories=['{}'])".format(configuration))
         return self._arm.send_joint_trajectory(configuration, timeout)
 
+    def send_goal(self, frameStamped, timeout=30, pre_grasp=False, first_joint_pos_only=False,
```
Comment (LarsJanssenTUe) [inline robot_skills/src/robot_skills/arms.py]: Shouldn't pre_grasp be 'true' by default?

File: robot_skills/src/robot_skills/hero.py [hunk 3]
```diff
@@ -40,7 +40,6 @@ def __init__(self, wait_services=False):
        
         #rename joint names
         self.parts['leftArm'].joint_names = self.parts['leftArm'].load_param('skills/arm/joint_names')
```
Comment (LarsJanssenTUe) [inline robot_skills/src/robot_skills/hero.py]: Shouldn't this also be generic?

File: robot_skills/src/robot_skills/robot.py [hunk 4]
```diff
@@ -70,7 +71,7 @@ def configure(self):
         # Wait for connections
         s = rospy.Time.now()
         for partname, bodypart in self.parts.iteritems():
-            bodypart.wait_for_connections(1.0)
+            bodypart.wait_for_connections(0.5)
```
Comment (LarsJanssenTUe) [inline robot_skills/src/robot_skills/robot.py]: Why is this changed?

File: robot_skills/src/robot_skills/robot.py [hunk 5]
```diff
@@ -194,16 +199,32 @@ def seq_or_none(obj):
                 for match in matches:
                     matching_grippers.update(match)
 
+            # ToDO: HACK for not specifying any requirements
```
Comment (LarsJanssenTUe) [inline robot_skills/src/robot_skills/robot.py]: Are there issues for these hacks?

File: robot_skills/src/robot_skills/robot.py [hunk 6]
```diff
@@ -184,6 +185,10 @@ def seq_or_none(obj):
 
             # Grippers
             matching_grippers = set()
+            # ToDO: HACK for not specifying any requirements
```
Comment (LoyVanBeek) [inline robot_skills/src/robot_skills/robot.py]: +100 for this. Makes API much easier to use

File: robot_smach_states/src/robot_smach_states/manipulation/grab.py [hunk 7]
```diff
@@ -275,29 +268,30 @@ def associate(self, original_entity):
...
-        if hasattr(userdata, 'arm'):
-            arm = self._robot.get_arm(required_gripper_types=arms.GripperTypes.GRASPING,
-                                      required_arm_name=userdata.arm.side)
-        else:
-            arm = None
-        self._robot.torso.reset()  # Move up to make resetting of the arm safer
+        arm = self.arm_designator.resolve()
+        arm.reset()
+
+        # ToDo: move up... This is AMIGO specific
```
Comment (LarsJanssenTUe) [inline robot_smach_states/src/robot_smach_states/manipulation/grab.py]: Isn't this an easy fix? If so please do it now!

File: robot_smach_states/src/robot_smach_states/navigation/navigate_to_grasp.py [hunk 8]
```diff
@@ -6,28 +6,29 @@
...
 
-        check_resolve_type(arm_designator, Arm) #Check that the arm_designator resolves to an Arm
+        check_resolve_type(arm_designator, PublicArm)  # Check that the arm_designator resolves to an Arm
+
         self.arm_designator = arm_designator
         if not arm_designator:
-            rospy.logerr('NavigateToGrasp: side should be determined by entity_designator. Please specify left or right, will default to left')
-            self.arm_designator = Designator(robot.leftArm)
+            rospy.logerr('NavigateToGrasp: side should be determined by entity_designator.\
+            Please specify left or right, will default to left. This is Deprecated')
```
Comment (LarsJanssenTUe) [inline robot_smach_states/src/robot_smach_states/navigation/navigate_to_grasp.py]: Why left?

File: robot_skills/src/robot_skills/arms.py (hunk 2 above)
Comment (MatthijsBurgh) [reply robot_skills/src/robot_skills/arms.py]: I don't know. Just matching the API of the hidden arm.

File: robot_skills/src/robot_skills/hero.py (hunk 3 above)
Comment (MatthijsBurgh) [reply robot_skills/src/robot_skills/hero.py]: Jep, but it isn't at the moment. So please make an issue for it.

File: robot_skills/src/robot_skills/robot.py (hunk 4 above)
Comment (MatthijsBurgh) [reply robot_skills/src/robot_skills/robot.py]: Just to speed it up.

File: robot_smach_states/src/robot_smach_states/navigation/navigate_to_grasp.py (hunk 8 above)
Comment (MatthijsBurgh) [reply robot_smach_states/src/robot_smach_states/navigation/navigate_to_grasp.py]: As it says. It is deprecated. I don't know what to put in the print. So just added the deprecated sentence.

Comment (MatthijsBurgh) [reply robot_smach_states/src/robot_smach_states/navigation/navigate_to_grasp.py]: If you know something better. Feel free to change it.

File: challenge_gpsr/src/gpsr.py (hunk 1 above)
Comment (MatthijsBurgh) [reply challenge_gpsr/src/gpsr.py]: This was wrong. If not testing, you are in a real challenge. So you're asking for confirmation.

File: robot_skills/src/robot_skills/robot.py (hunk 5 above)
Comment (MatthijsBurgh) [reply robot_skills/src/robot_skills/robot.py]: #767

File: robot_smach_states/src/robot_smach_states/manipulation/grab.py (hunk 7 above)
Comment (MatthijsBurgh) [reply robot_smach_states/src/robot_smach_states/manipulation/grab.py]: Easy hack...

File: robot_skills/src/robot_skills/robot.py (hunk 6 above)
Comment (alberth) [reply robot_skills/src/robot_skills/robot.py]: Except you can't have a guarantee that you have sufficient capabilities if you don't specify what you need to have.

Comment (MatthijsBurgh) [reply robot_skills/src/robot_skills/robot.py]: I know, but It would be too much to change at the moment.

File: robot_smach_states/src/robot_smach_states/manipulation/grab.py [hunk 9]
```diff
@@ -96,7 +99,6 @@ def execute(self, userdata):
         if not arm:
             rospy.logerr("Could not resolve arm")
             return "failed"
-        userdata.arm = arm.side  # Using userdata makes sure we don't need to do any more arm entity magic
```
Comment (jlunenburg) [inline robot_smach_states/src/robot_smach_states/manipulation/grab.py]: So how do you know that the arm you used for pre-grasping is also the arm used for grasping? This is not checked in any case so you´ve made it much less robust IMHO. If you don't like the user data stuff you must somehow ensure that the designator is locked so that it *can't go wrong*. That does make the designator stuff much more difficult so I don't understand why you reverted this.

File: robot_smach_states/src/robot_smach_states/navigation/navigate_to_grasp.py [hunk 10]
```diff
@@ -6,28 +6,29 @@
 import rospy
 
 # TU/e Robotics
-from cb_planner_msgs_srvs.srv import *
 from cb_planner_msgs_srvs.msg import *
-from robot_skills.arms import Arm
+from robot_skills.arms import PublicArm
```
Comment (jlunenburg) [inline robot_smach_states/src/robot_smach_states/navigation/navigate_to_grasp.py]: Personally, I think it would have been much nicer if the arm you import was called `Arm` and the other one `PrivateArm` (so the other way around).

**Comment by alberth** [comment]
On a general note, wouldn't it be useful to not merge competition stuff into master, but use some 'competition' branch instead? That would keep regular development and last minute hacking separated, so one doesn't have to untangle the mess from master afterwards.

File: robot_smach_states/src/robot_smach_states/manipulation/grab.py (hunk 9 above)
Comment (MatthijsBurgh) [reply robot_smach_states/src/robot_smach_states/manipulation/grab.py]: This was also not the case before.

**Comment by MatthijsBurgh** [comment]
> On a general note, wouldn't it be useful to not merge competition stuff into master, but use some 'competition' branch instead? That would keep regular development and last minute hacking separated, so one doesn't have to untangle the mess from master afterwards.

That does happen on tournaments. But right now the master was broken. Something that shouldn't have happened. So the hacking is not really related to the tournament, but to a broken master and limited time.

File: robot_smach_states/src/robot_smach_states/manipulation/grab.py (hunk 9 above)
Comment (jlunenburg) [reply robot_smach_states/src/robot_smach_states/manipulation/grab.py]: For sure it was for the reset (see changes between lines 290 and 300). If now for a robot with two arms the first arm gives an error in grasping, it might resolve to the second arm and reset that one. You now reverted it to the old situation that has proven to work very poorly... There was even a comment on this in the code....

Comment (MatthijsBurgh) [reply robot_smach_states/src/robot_smach_states/manipulation/grab.py]: Indeed for the reset it was. But it is better to do the locking.

**Comment by alberth** [comment]
Limited time doesn't count on master, it would count on a tournament branch, so as far as I am concerned you're doing tournament hacking in master now.
